security/src/main/java/io/micronaut/security/authentication/BlockingAuthenticationProviderFactory.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2017-2023 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.security.authentication;
+
+import io.micronaut.context.annotation.EachBean;
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.scheduling.TaskExecutors;
+import jakarta.inject.Named;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import reactor.core.scheduler.Schedulers;
+
+import java.util.concurrent.ExecutorService;
+
+/**
+ * A factory for adapting {@link BlockingAuthenticationProvider} beans to expose them as {@link AuthenticationProvider}.
+ *
+ * @since 4.5.0
+ */
+@Factory
+@Internal
+class BlockingAuthenticationProviderFactory {
```

Comment:
Can we not call it blocking? It doesn't need to block

Reply:
I have changed the API `ImperativeAuthenticationProvider`. 

If the user implementation is blocking he should add the `@Blocking` annotation to the overridden method.

Reply:
What is the plan? Rename it in v5? Or keep it like this forever?
IMHO I would create a new API non-reactive `AuthenticationProvider` and `ReactiveAuthenticationProvider` in a different package and deprecate the old one.

Reply:
If we were starting from zero, I would not be against  the naming to be `ReactiveAuthenticationProvider` for the publisher api and `AuthenticationProvider` for the imperative API.

But we already have an API in place, which most users will have implemented in their applications and I am not keen to forcing them to change their code for a better name.

Having the same API name in different packages sounds tricky to me as well. 

I don't think having two APIs: the existing `AuthenticationProvider` which is reactive and a new `ImperativeAuthenticationProvider` which is not reactive is too bad.

Reply:
@dstepanov I have deprecated

io.micronaut.security.authentication.AuthenticationProvider

and created two APIs: 

- Imperative => `io.micronaut.security.authentication.provider.AuthenticationProvider` 
- Reactive => `io.micronaut.security.authentication.provider.ReactiveAuthenticationProvider`

https://github.com/micronaut-projects/micronaut-security/pull/1526/commits/c795eaada8c7253f0a82eef4334195a575273eb8

Reply:
@sdelamo I like this - it's clean and more consistent with the terminology we use elsewhere (i.e., Micronaut Data).